### DISCUSSION TITLE
Http response time in format=prometheus

### DISCUSSION THREAD

**Discussion by ehymui** [javamelody/javamelody#1208] [General]
Hello,

I am new to JavaMelody and struggling on how to reveal the metrics for http mean and max time when the format is prometheus.  Anyone has idea ?

**Comment by evernat** [comment]
Of course, you have read https://github.com/javamelody/javamelody/wiki/UserGuideAdvanced#exposing-metrics-to-prometheus

If you look at the example http://javamelody.org/demo/monitoring?format=prometheus, you will see two Prometheus metrics among others :
- javamelody_http_hits_count
- javamelody_http_duration_millis

Those two metrics are of type "counter" (the values always increase over time, since the start of the monitored app).

To have a useful http mean response time or a max time, you need to decide on which period you want a value (that is probably not from the start of the app): last hour ? or last minute ?
Then you may want a graph of that value over time.

If I make no mistake, you can get a graph over time of http mean response time for each minute using PromQL and a division operator with something like that : 
`increase(javamelody_http_duration_millis{instance="localhost:8080", job="my_app"}[1m]) / increase(javamelody_http_hits_count{instance="localhost:8080", job="my_app"}[1m])`

The http max response time is not available at the moment.

Alternatively, you could get values from graphs displayed in the monitoring page (beware that it is slower because it needs to read values from RRD files). For that use a URL like http://javamelody.org/demo/monitoring?format=prometheus&includeLastValue=true
That is, put a `includeLastValue` param like that in your scrape config:
```
    params:
      format: ['prometheus']
      includeLastValue: ['true']
```
With `includeLastValue`, you will have new metrics exported, including `javamelody_last_value_http_mean_times`, which is a gauge type metric and which is directly what you want with a 1 minute resolution. The metrics in the javamelody graphs have a new value each minute per default, even if you scrape it more often than once per minute. Again the http max response time is not available for Prometheus from those metrics.

**Reply by evernat** [reply]
(oups) inverted the numerator and denominator in the division above

**Reply by ehymui** [reply]
Thank you very much for your explanation!
